core/src/main/java/jenkins/util/SystemProperties.java

```diff
@@ -0,0 +1,264 @@
+     * 
+     * @param      key   the name of the system property.
+     * @return     the string value of the system property,
+     *             or {@code null} if there is no property with that key.
+     *
+     * @exception  NullPointerException if {@code key} is {@code null}.
+     * @exception  IllegalArgumentException if {@code key} is empty.
+     */
+    @CheckForNull
+    public static String getString(String key) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if (value != null) {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Property (system): {0} => {1}", new Object[] {key, value});
+            }
+            return value;
+        }
+        
+        value = tryGetValueFromContext(key);
+        if (value != null) {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Property (context): {0} => {1}", new Object[]{key, value});
+            }
+            return value;
+        }
+        
+        if (LOGGER.isLoggable(Level.CONFIG)) {
+            LOGGER.log(Level.CONFIG, "Property (not found): {0} => {1}", new Object[] {key, value});
+        }
+        return null;
+    }
+
+    /**
+     * Gets the system property indicated by the specified key, or a default value.
+     * This behaves just like {@link System#getProperty(java.lang.String, java.lang.String)}, except
+     * that it also consults the {@link ServletContext}'s "init" parameters.
+     * {@link ServletContext} check will be skipped if the context is not initialized.
+     * 
+     * @param      key   the name of the system property.
+     * @param      def   a default value.
+     * @return     the string value of the system property,
+     *             or {@code null} if there is no property with that key.
+     *
+     * @exception  NullPointerException if {@code key} is {@code null}.
+     * @exception  IllegalArgumentException if {@code key} is empty.
+     */
+    @CheckForNull
+    public static String getString(String key, @CheckForNull String def) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if (value != null) {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Property (system): {0} => {1}", new Object[] {key, value});
+            }
+            return value;
+        } 
+        
+        value = tryGetValueFromContext(key);
+        if (value != null) {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Property (context): {0} => {1}", new Object[]{key, value});
+            }
+            return value;
+        }
+        
+        value = def;
+        if (LOGGER.isLoggable(Level.CONFIG)) {
+            LOGGER.log(Level.CONFIG, "Property (default): {0} => {1}", new Object[] {key, value});
+        }
+        return value;
+    }
```

Comment:
NIT: some repetition of code, could have re-used the above `getString(String key)`, but that's okey it's a NIT